File: example/novel/novel_test.go
package novel_test

import (
	"testing"

	"github.com/ditrit/specimen/example/novel"
	"github.com/ditrit/specimen/go/specimen"
)

func deserialize_book(book_data specimen.Dict) novel.Book {
	return novel.Book{
		Title:    book_data["title"].(string),
		LeftPage: book_data["left_page"].(int),
		Size:     book_data["size"].(int),
	}
}

var codeboxSet = specimen.MakeCodeboxSet(map[string]specimen.BoxFunction{
	"turn_page": func(s *specimen.S, input specimen.Dict) {
		book := deserialize_book(input["book"].(specimen.Dict))
		book.TurnPage(input["turn_page_count"].(int))

		s.ExpectEqual(
			specimen.Dict{
				"title":     book.Title,
				"left_page": book.LeftPage,
				"size":      book.Size,
			},
			input["expected_result"].(specimen.Dict),
			"result comparison",
		)
	},

	"turn_page_expect_page": func(s *specimen.S, input specimen.Dict) {
		book := deserialize_book(input["book"].(specimen.Dict))
		book.TurnPage(input["turn_page_count"].(int))

		if input["expected_left_page"] != book.LeftPage {
			s.Fail("page mismatch")
		}
	},

	"get_page": func(s *specimen.S, input specimen.Dict) {
		book := deserialize_book(input["book"].(specimen.Dict))
		s.ExpectEqual(
			book.GetPage(),
			input["expected_result"].(int),
			"result comparison",
		)
	},
})

func TestNovel(t *testing.T) {
	specimen.Run(
		t,
		codeboxSet,
		[]specimen.File{
			specimen.ReadLocalFile("novel_data.yaml"),
			specimen.ReadLocalFile("novel_data_with_alias.yaml"),
		},
	)
}

File: rust/tests/novel_test.rs
use serde::Deserialize;

#[derive(Debug, Default, Deserialize)]
struct Book {
    left_page: i32,
    size: i32,
}

impl Book {
    fn turn_page(&mut self, count: i32) {
        self.left_page += 2 * count;

        if self.left_page < 0 {
            self.left_page = 0;
        } else if self.left_page >= self.size {
            self.left_page = self.size - 1;
        }
    }

    fn get_page(&self) -> i32 {
        self.left_page
    }
}

fn deserialize_book(data: &Box<str>) -> Book {
    serde_yaml::from_str(data).unwrap()
}

fn turn_page(input: &specimen::Dict) -> Result<(), Box<str>> {
    let mut book = deserialize_book(&input["book"]);
    let count = input["turn_page_count"].parse().unwrap();
    book.turn_page(count);
    Ok(())
}

fn turn_page_expect_page(input: &specimen::Dict) -> Result<(), Box<str>> {
    let mut book = deserialize_book(&input["book"]);
    let count = input["turn_page_count"].parse().unwrap();
    book.turn_page(count);
    let expected_page = input["expected_left_page"].parse().unwrap();
    match book.left_page == expected_page {
        true => Ok(()),
        false => Err(Box::from(format!(
            "Expected page: {}, Actual page: {}",
            expected_page, book.left_page
        ))),
    }
}

fn get_page(input: &specimen::Dict) -> Result<(), Box<str>> {
    let book = deserialize_book(&input["book"]);
    let expected = input["expected_result"].parse().unwrap();
    match book.get_page() == expected {
        true => Ok(()),
        false => Err(Box::from(format!(
            "Expected page: {}, Actual page: {}",
            book.get_page(),
            book.left_page
        ))),
    }
}

#[test]
fn test_novel() {
    specimen::run(
        &mut |tile: &specimen::Dict| -> Result<(), Box<str>> {
            match &*tile["box"] {
                "get_page" => get_page(tile),
                "turn_page_expect_page" => turn_page_expect_page(tile),
                "turn_page" => turn_page(tile),
                _ => Err(Box::from(format!("Unknown box: {}", tile["box"]))),
            }
        },
        &[specimen::file::File::read_local_file(
            "../test/novel/novel_data.yaml",
        )],
    )
}
